### PR TITLE
feat(simulation): initial simulation visualisation setup

### PR DESCRIPTION
**Pull request by Matt-Hayter**
First steps in creating visualisation. Bar chart visuals are as desired, and simulation runs for predator-prey. Need to configure params and time steps to get it working smoothly.

### CODE REVIEW

File: backend/biological_models_app/pred_prey/simulation.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                break
+            #If peak hasn't been found before t_max_stop, end sim
+            if t >= self.t_max_stop:
+                return
+
+        self.t_axis = np.linspace(0, len(self.N)*self.dt, len(self.N))
+            
+        #Continue iterations now first trough is found
+        trough_counter = 1
+        while True:
```
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py]: While True not good coding practice convert to `while trough_counter < self.max_troughs:`

File: backend/biological_models_app/pred_prey/simulation.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def run_sim(self):
+        """
+        Run Euler iterations for each time step until max_t
+        """
+        self.N.append(self.N_0)
+        self.P.append(self.P_0)
+        
+        t = 0
+        #Perform Euler until first prey trough is found
+        while True:
```
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py:58]:
```suggestion
        while t < self.t_max_stop
```

File: backend/biological_models_app/pred_prey/simulation.py (hunk 1 above)
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py]:
```suggestion
        while trough_counter < self.max_troughs:
```

File: backend/biological_models_app/pred_prey/simulation.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            t += self.dt
+            self.Euler_method()
+            if self.dNdt[-2] < 0 and self.dNdt[-1] >= 0:
+                break
+            #If peak hasn't been found before t_max_stop, end sim
+            if t >= self.t_max_stop:
+                return
```
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py:65]:
```suggestion
```

File: backend/biological_models_app/pred_prey/simulation.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            
+        #Continue iterations now first trough is found
+        trough_counter = 1
+        while True:
+            t += self.dt
+            self.Euler_method()
+            if self.dNdt[-2] < 0 and self.dNdt[-1] >= 0: #Count troughs as they are passed
+                trough_counter += 1
+            if trough_counter == self.max_troughs: #End simulation when desired through is found
+                return
```
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py]:
```suggestion
```

File: backend/biological_models_app/pred_prey/simulation.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        self.N.append(self.N_0)
+        self.P.append(self.P_0)
+        
+        t = 0
+        #Perform Euler until first prey trough is found
+        while True:
+            t += self.dt
+            self.Euler_method()
+            if self.dNdt[-2] < 0 and self.dNdt[-1] >= 0:
+                break
```
Comment (TJ-io) [inline backend/biological_models_app/pred_prey/simulation.py:62]: Are you sure you want break here and not return?

File: frontend/frontend/src/components/Common/RacerBarChart.vue [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    simMaxVal: Number
+  },
+  data() {
+    return {
+      racerChart: null
+    };
+  },
+  watch: {
+    initialConditions: function(value) { //Update initial bar lengths upon user param change
+      for(let i = 0; i < value.length; i++) {
```
Comment (TJ-io) [inline frontend/frontend/src/components/Common/RacerBarChart.vue]:
```suggestion
    initialConditions: function(dataPoints) { //Update initial bar lengths upon user param change
      for(let i = 0; i < dataPoints.length; i++) {
```

Use a descriptive name

File: frontend/frontend/src/components/Common/RacerBarChart.vue [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        this.chartConfig.data.datasets[0].data[i] = this.initialConditions[i]
+      }
+      this.racerChart.update();
+    },
+    simRunning: function(value) {
```
Comment (TJ-io) [inline frontend/frontend/src/components/Common/RacerBarChart.vue]: Use a descriptive name not value

File: frontend/frontend/src/components/Common/RacerBarChart.vue [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      this.chartConfig.options.scales.x.max = this.simMaxVal //Resize bar plot to fit sim
+      this.racerChart.update();
+      //Visualise simulation if simRunning turns to true
+      if (value) {
+        let step = 0 //For all simulation time steps
+        setInterval(() => {
+          //Set each chart data element to it's corresponding simulation data index, on each iteration
+          for (let i = 0; i < this.simData.length; i++)
+          this.chartConfig.data.datasets[0].data[i] = this.simData[i][step]
+          this.racerChart.update();
```
Comment (TJ-io) [inline frontend/frontend/src/components/Common/RacerBarChart.vue]:
```suggestion
          for (let i = 0; i < this.simData.length; i++) {
              this.chartConfig.data.datasets[0].data[i] = this.simData[i][step]
          }
          this.racerChart.update();
```

File: frontend/frontend/src/components/ConfigBar/components/PresetButtons.vue [hunk 9]
```diff
@@ -96,18 +100,10 @@ export default {
       return this.userPresets
     },
     emptyPresetsSignedIn() {
-      if (this.userPresets.length == 0 && this.$store.state.activeUser.isActive) {
-        return true
-      } else {
-        return false
-      }
+      return this.userPresets.length == 0 && this.$store.state.activeUser.isActive
```
Comment (TJ-io) [inline frontend/frontend/src/components/ConfigBar/components/PresetButtons.vue]:
```suggestion
      return !this.userPresets.length && this.$store.state.activeUser.isActive
```

File: frontend/frontend/src/components/ConfigBar/components/PresetButtons.vue [hunk 10]
```diff
@@ -96,18 +100,10 @@ export default {
...
     },
     emptyPresetsSignedOut() {
-      if (this.userPresets.length == 0 && !this.$store.state.activeUser.isActive) {
-        return true
-      } else {
-        return false
-      }
+      return this.userPresets.length == 0 && !this.$store.state.activeUser.isActive
```
Comment (TJ-io) [inline frontend/frontend/src/components/ConfigBar/components/PresetButtons.vue]:
```suggestion
      return !this.userPresets.length && !this.$store.state.activeUser.isActive
```

File: frontend/frontend/src/views/PredatorPrey/PredPreyChartConfig.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+//Contains configuration for racer bar chart
+let chartData = {
```
Comment (TJ-io) [inline frontend/frontend/src/views/PredatorPrey/PredPreyChartConfig.js:2]:
```suggestion
const CHART_DATA= {
```

if the value doesn't change and is a constant

**Review by TJ-io — CHANGES_REQUESTED** [review]
Couple changes but looking good!

File: backend/biological_models_app/pred_prey/simulation.py (hunk 1 above)
Comment (TJ-io) [reply backend/biological_models_app/pred_prey/simulation.py]: Try do while - [see examples here](https://www.freecodecamp.org/news/python-do-while-loop-example/)

File: backend/biological_models_app/pred_prey/simulation.py (hunk 2 above)
Comment (Matt-Hayter) [reply backend/biological_models_app/pred_prey/simulation.py:58]: Indefinite loop is prevented using 
if t >= self.t_max_stop: return
so I'll leave this one as discussed.

File: backend/biological_models_app/pred_prey/simulation.py (hunk 5 above)
Comment (Matt-Hayter) [reply backend/biological_models_app/pred_prey/simulation.py:62]: Yeah, means the following while loop will be performed sequentially

File: frontend/frontend/src/views/PredatorPrey/PredPreyChartConfig.js (hunk 11 above)
Comment (Matt-Hayter) [reply frontend/frontend/src/views/PredatorPrey/PredPreyChartConfig.js:2]: This one does actually have keys that change, from within RacerBarChart.vue

**Review by TJ-io — APPROVED** [review]

**Pull request merged**